feat(dropdown-field): implement combobox for dropdown fields

orbitalsqwib opened the pull request:
## Problem

Currently, the dropdown fields in CheckFirst do not have search capabilities. This is bad UX as users will have to scan long lists of options manually to find their desired option.

Closes #425

## Solution

This PR adds a reusable and performant combobox component to replace the current `select` dropdown for dropdown fields.

**Features**:

- Fast and snappy search behaviour
- Supports large lists, e.g > 1000 rows

## Before & After Screenshots

**BEFORE**:
<img width="531" alt="Screenshot 2021-04-30 at 3 28 51 PM" src="https://user-images.githubusercontent.com/21305518/116662815-d52d4800-a9c8-11eb-8b21-605aa4aacc90.png">

**AFTER**:
<img width="530" alt="Screenshot 2021-04-30 at 3 28 39 PM" src="https://user-images.githubusercontent.com/21305518/116662781-ca72b300-a9c8-11eb-87de-1938948160b6.png">

## Tests

- [x] Check that search works
- [x] Check that clicking out of the input field resets the input field value to the last selected element
- [x] Check that clicking a row via mouse or arrow keys+enter sets the input field value to that row's label
- [x] Check that the correct value is returned by the combobox when referenced in a result field

## Deploy Notes

**New dependencies**:

- `react-window` : used to reduce the render lag on very large lists

**New dev dependencies**:

- `@types/react-window` : used to type `react-window`

Commit:
feat(dropdown-field): add search to dropdown

- implement new search dropdown component to replace select

Commit:
feat(dropdown-field): enhance efficiency and design

- implemented react-window to reduce performance impact of downshift event
re-renders
- implemented dynamic dropdown/dropup behaviour
- renamed search-dropdown component to combobox
- implemented ui enhancements including clear button and decorator icons
- generalised combobox behaviour for easier reuse in other files.

Commit:
fix(dropdown-field): improve accessibility and fix minor bugs

- add input description for combobox
- improve dropdown height calculation
- ensure that menu is always displayed to avoid mounting and unmounting
the component, which could result in accessibility issues

Commit:
fix(package-json): shift react-window types to dev dependencies

lamkeewei commented on src/client/components/common/Combobox.tsx:
```diff
@@ -0,0 +1,358 @@
+                  topDropdownInset: dropdownInset,
+                } as ItemRendererProps['data']
+              }
+              innerElementType={innerElementType}
```
Do we need to set this prop? Isn't the default value a `div` and also the styling is applied on the `ListItem`?

orbitalsqwib replied:
This prop is required to add the 8px in-list padding at the top and bottom as required in the design, as the row elements use absolute positioning within the div and thus require us to manually extend the height of the div element to account for the padding.

lamkeewei commented on src/client/components/common/Combobox.tsx:
```diff
@@ -0,0 +1,358 @@
+            ...props.style,
+            // purely for adding padding insets to FixedStyleList
+            // see https://codesandbox.io/s/react-window-list-padding-dg0pq
+            top: `${parseFloat(`${props.style.top}`) + topDropdownInset}px`,
```
```suggestion
            top: `${parseFloat(`${props.style.top}`) + topDropdownInset}px`,
            listStyle: 'none',
```

Currently bullets are still shown for list items.

<img src='https://user-images.githubusercontent.com/3666479/116958543-9b608800-accd-11eb-8410-86a5da1526dd.png' width=500 />

lamkeewei commented on src/client/components/common/Combobox.tsx:
```diff
@@ -0,0 +1,358 @@
+  const inputRef = inputOptions?.forwardRef || fallbackInputRef
+
+  const dropdownList = useRef<FixedSizeList & HTMLElement>(null)
+  const searchResults = useRef<ComboboxItem[]>(items)
```
Is there as specific reason to use `useRef` instead of `useState`? 

Generally, `useRef` should be used for things that persists throughout the entire lifecycle of a component and `useState` is used for things that causes a re-render. Given that when the search results changes, a re-render would be necessary, might be more appropriate to use `useState`.

lamkeewei commented on src/client/components/common/Combobox.tsx:
```diff
@@ -0,0 +1,358 @@
+  const getTopSearchResults = (inputValue: string) => {
+    let searchResults: ComboboxItem[]
+    if (inputValue === '') {
+      searchResults = items
```
I think the [default behaviour](https://runkit.com/embed/gzr7mj0wkam1) for `matchSorter` is to return all items when `inputValue` is empty.

lamkeewei reviewed (changes requested):
Left some comments/questions! Two other general comments are:
1. Might want to look into the use of [useMultiStyleConfig](https://chakra-ui.com/docs/theming/component-style#styling-multipart-components) to centralise styles for the different parts of the combo box (input, dropdown item, etc).
2. Not sure if it might be worth refactoring to make use of `useCombobox` to simplify the code a little.

Great job on looking into the use of `react-window` to handle potentially huge lists.

Commit:
refactor(combobox): clean up code, fix list styles

- use use-state instead of use-ref for values that affect re-renders
- use list-styles: none for list items to fix dropdown formatting
- renamed dropdown-list to dropdown-list-ref to highlight that it is a
  ref to a element
- removed unnecessary empty string check for match-sorter

orbitalsqwib commented:
Hi @lamkeewei, thanks for the helpful comments on this PR!

In regards to 1, I think I'll open a new issue to implement the style centralization and work on it on a separate PR. 

However, I think 2 might not be worth the refactor effort as it will affect the order in which the functions are called, thus essentially requiring large changes to the current code. Do let me know if this is something you feel strongly about and I'll try my best to implement it.

lamkeewei reviewed (approved):
lgtm!
